### PR TITLE
SQL: rewrite ROUND and TRUNCATE functions with a different optional parameter handling method

### LINKED ISSUE #40001: SQL: GROUP BY with function group returns error
This following statement works as expected: 
```select convert(truncate(extract(YEAR from "timestamp")), SQL_BIGINT) as yr from "kibana_sample_data_flights" group by yr;```

When the `yr` is swapped with the group expression it replaces, an exception is raised:
```
sql> select convert(truncate(extract(YEAR from "timestamp")), SQL_BIGINT) as yr from "kibana_sample_data_flights" group by convert(truncate(extract(YEAR from "timestamp")), SQL_BIGINT);
Server error [Server sent bad type [folding_exception]. Original type was [line 1:8: Cannot find grouping for 'convert(truncate(extract(YEAR from "timestamp")), SQL_BIGINT)']. [FoldingException[line 1:8: Cannot find grouping for 'convert(truncate(extract(YEAR from "timestamp")), SQL_BIGINT)']
        at org.elasticsearch.xpack.sql.planner.QueryFolder$FoldAggregate.rule(QueryFolder.java:321)
        at org.elasticsearch.xpack.sql.planner.QueryFolder$FoldAggregate.rule(QueryFolder.java:206)
        at org.elasticsearch.xpack.sql.tree.Node.lambda$transformUp$11(Node.java:196)
        at org.elasticsearch.xpack.sql.tree.Node.transformUp(Node.java:190)
        at org.elasticsearch.xpack.sql.tree.Node.lambda$transformUp$10(Node.java:188)
        at org.elasticsearch.xpack.sql.tree.Node.transformChildren(Node.java:209)
        at org.elasticsearch.xpack.sql.tree.Node.transformUp(Node.java:188)
        at org.elasticsearch.xpack.sql.tree.Node.transformUp(Node.java:196)
        at org.elasticsearch.xpack.sql.planner.QueryFolder$FoldingRule.apply(QueryFolder.java:589)
        at org.elasticsearch.xpack.sql.planner.QueryFolder$FoldingRule.apply(QueryFolder.java:585)
        at org.elasticsearch.xpack.sql.rule.RuleExecutor$Transformation.<init>(RuleExecutor.java:82)
        at org.elasticsearch.xpack.sql.rule.RuleExecutor.executeWithInfo(RuleExecutor.java:158)
        at org.elasticsearch.xpack.sql.rule.RuleExecutor.execute(RuleExecutor.java:130)
        at org.elasticsearch.xpack.sql.planner.QueryFolder.fold(QueryFolder.java:85)
        at org.elasticsearch.xpack.sql.planner.Planner.foldPlan(Planner.java:38)
        at org.elasticsearch.xpack.sql.planner.Planner.plan(Planner.java:28)
        at org.elasticsearch.xpack.sql.session.SqlSession.lambda$physicalPlan$4(SqlSession.java:156)
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:62)
        at org.elasticsearch.xpack.sql.session.SqlSession.lambda$optimizedPlan$3(SqlSession.java:152)
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:62)
        at org.elasticsearch.xpack.sql.session.SqlSession.lambda$preAnalyze$2(SqlSession.java:140)
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:62)
        at org.elasticsearch.xpack.sql.analysis.index.IndexResolver.lambda$resolveAsMergedMapping$3(IndexResolver.java:245)
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:62)
        at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:68)
        at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:64)
        at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43)
        at org.elasticsearch.action.fieldcaps.TransportFieldCapabilitiesAction.lambda$doExecute$1(TransportFieldCapabilitiesAction.java:84)
        at org.elasticsearch.action.fieldcaps.TransportFieldCapabilitiesAction$1.onResponse(TransportFieldCapabilitiesAction.java:97)
        at org.elasticsearch.action.fieldcaps.TransportFieldCapabilitiesAction$1.onResponse(TransportFieldCapabilitiesAction.java:93)
        at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:68)
        at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:64)
        at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43)
        at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:271)
        at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:255)
        at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1108)
        at org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:1189)
        at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1169)
        at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:54)
        at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:47)
        at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:30)
        at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$1.doRun(TransportSingleShardAction.java:117)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
]]
```

### PR DESCRIPTION
**Pull request by astefan**
The optional parameter for these two functions was handled in the constructor by passing the value of `0` whenever the parameter was missing. This had an apparently minor implication of defining a second child for the function (the second parameter). If the value was kept as `null` (missing) the function would have had a single child (the first parameter).

This affected the comparison operation from [the Analyzer](https://github.com/elastic/elasticsearch/blob/7.x/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java#L896) and the same function that was already resolved wouldn't match the new function found because the new one had no second parameter while the resolved one had the parameter `0`. Because of this comparison failing, a `GROUP BY` the same function would fail.

This PR also fixes a bug with ATAN2 and POWER functions which didn't have any Painless methods.

Fixes https://github.com/elastic/elasticsearch/issues/40001.

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isNumeric(left, sourceText(), ParamOrdinal.FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+
+        }
+
+        return right == null ? TypeResolution.TYPE_RESOLVED : isNumeric(right, sourceText(), ParamOrdinal.SECOND);
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java]: The `right` should be of integer type (not rational), so we should catch it here before the execution time.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/Truncate.java [hunk 2]
```diff
@@ -6,36 +6,42 @@
...
+    protected BinaryOptionalMathOperation operation() {
+        return BinaryOptionalMathOperation.TRUNCATE;
     }
-
+    
     @Override
-    public DataType dataType() {
-        return left().dataType();
+    public Expression replaceChildren(List<Expression> newChildren) {
+        if (right() != null && newChildren.size() != 2) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/Truncate.java]: I'd extract this to the parent method and leave here only the instance creation, as it's duplicated also in `Round`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    @Override
+    public ScriptTemplate asScript() {
+        ScriptTemplate leftScript = asScript(left);
+        ScriptTemplate rightScript = right == null ? null : asScript(right);
+
+        return asScriptFrom(leftScript, rightScript);
+    }
+
+    private ScriptTemplate asScriptFrom(ScriptTemplate leftScript, ScriptTemplate rightScript) {
+        if (right == null) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java]: Wouldn't be possible to avoid the "double" methods in painless and just have one with both args where null is passed for the right arg?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalMathProcessor.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.sql.SqlIllegalArgumentException;
+import org.elasticsearch.xpack.sql.expression.gen.processor.Processor;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.function.BiFunction;
+
+/**
+ * Processor for binary mathematical operations that can have the second parameter optional.
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalMathProcessor.java]: Maybe rephrase: ... that have a second optional parameter.

File: x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec [hunk 5]
```diff
@@ -200,6 +200,23 @@ ROUND(SQRT(CAST(EXP(languages) AS SMALLINT)),2):d|  COUNT(*):l
 null                                             |10
 ;
 
+groupByRoundWithTwoParams
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.csv-spec]: Can't this be tested in QueryTranslatorTests?

**Review by matriv — COMMENTED** [review]
Nice refactoring. Left some comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java]: Gave it a try, but failed. If you have any ideas, I'd love to hear them. Otherwise, I'll try later today or tomorrow.

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/default-distro

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java (hunk 3 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java]: I took some time and it works like that: https://github.com/matriv/elasticsearch/commit/b35090e91f54f7db00e916e4671ad00fcd551cf4

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/default-distro

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/math/BinaryOptionalNumericFunction.java]: Indeed. Nice one. I've incorporated it in the PR. Thanks.

**Review by matriv — APPROVED** [review]
LGTM

**Review by costin — APPROVED** [review]
Considering the timing and everything, I'm fine with the approach.
Going forward I think we need a more generic way of dealing with optional arguments - whether that requires updates in the Analyzer or/and the functions remains to be determined.
